File: AccuPay/Domain/Payroll/ReleaseThirteenthMonthPay.vb
Option Strict On

Imports AccuPay.DB
Imports AccuPay.Entity
Imports AccuPay.Utils
Imports Microsoft.EntityFrameworkCore

Public Class ReleaseThirteenthMonthPay

    Private thirteenthMonthPayname As String = "13th Month Pay"
    Private adjustmentCategoryName As String = "Adjustment Type"

    Public Sub New(dateFrom As Date, dateTo As Date, currentPeriodID As Integer?)
        Dim data = GetThirteenthMonthPays(dateFrom, dateTo)

        Using context = New PayrollContext()
            Dim paystubs =
                (From p In context.Paystubs.Include(Function(p) p.ActualAdjustments)
                 Where p.PayPeriodID.Value = currentPeriodID.Value).
                 ToList()

            Dim product =
                (From p In context.Products
                 Where p.OrganizationID.Value = z_OrganizationID And
                     p.PartNo = thirteenthMonthPayname And
                     p.Category = adjustmentCategoryName).
                 FirstOrDefault()

            Dim adjustmentCategoryId =
                context.Products.
                    Where(Function(p) CBool(p.OrganizationID.Value = z_OrganizationID _
                            And p.Category = adjustmentCategoryName)).
                        FirstOrDefault.
                            CategoryID

            'Provide this kind of Product if doesn't exists
            If product Is Nothing Then
                Dim thirteentMonthPayProduct = New Product() With {
                .OrganizationID = z_OrganizationID,
                .Created = Now.Date,
                .CreatedBy = z_User,
                .LastUpd = Now.Date,
                .LastUpdBy = z_User,
                .PartNo = thirteenthMonthPayname,
                .Name = thirteenthMonthPayname,
                .Description = String.Empty,
                .Category = adjustmentCategoryName,
                .CategoryID = adjustmentCategoryId.Value
                }
                Dim dfsdf = context.Products.Add(thirteentMonthPayProduct)
                context.SaveChanges()

                product = thirteentMonthPayProduct
            End If

            For Each paystub In paystubs
                'Preserve the existing ActualAdjustments, except the 13th month pay
                Dim non13thMonthAdjs = paystub.ActualAdjustments.Where(Function(aa) aa.ProductID.Value <> product.RowID.Value)

                'Reset the ActualAdjustments
                paystub.ActualAdjustments.Clear()

                'Restore the preserved ActualAdjustments
                If non13thMonthAdjs.Any() Then
                    For Each aa In non13thMonthAdjs
                        paystub.ActualAdjustments.Add(aa)
                    Next
                End If

                ' Get first the total thirteenth month pay that the employee will receive.
                Dim row = data.
                    Select($"EmployeeID = '{paystub.EmployeeID}'").
                    FirstOrDefault()
                Dim thirteenthMonthPay = CDec(row("ThirteenthMonthPay"))

                ' Add the new thirteenth pay as an adjustment to the paystub
                paystub.ActualAdjustments.Add(New ActualAdjustment() With {
                    .ProductID = product.RowID,
                    .OrganizationID = z_OrganizationID,
                    .CreatedBy = z_User,
                    .PayAmount = AccuMath.CommercialRound(thirteenthMonthPay)
                })
            Next

            context.SaveChanges()
        End Using
    End Sub

    Private Function GetThirteenthMonthPays(dateFrom As Date, dateTo As Date) As DataTable
        Dim thirteenthMonthPay = $"
            SELECT pst.EmployeeID, SUM(tmp.Amount) AS ThirteenthMonthPay
            FROM thirteenthmonthpay tmp
            INNER JOIN paystub pst
                    ON pst.RowID = tmp.PaystubID
            WHERE @PayFromDate <= pst.PayFromDate AND
                  pst.PayToDate <= @PayToDate AND
                  pst.OrganizationID = @OrganizationID
            GROUP BY pst.EmployeeID;"

        Dim params = New Dictionary(Of String, Object) From {
            {"@OrganizationID", z_OrganizationID},
            {"@PayFromDate", dateFrom},
            {"@PayToDate", dateTo}
        }

        Dim query = New SqlToDataTable(thirteenthMonthPay, params)
        Return query.Read()
    End Function

End Class

File: Accupay.Data/Services/ReleaseThirteenthMonthPay.cs
using AccuPay.Data.Entities;
using AccuPay.Utilities;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class ReleaseThirteenthMonthPay
    {
        private readonly int _organizationId;
        private readonly int _userId;

        private const string thirteenthMonthPayname = "13th Month Pay";
        private const string adjustmentCategoryName = "Adjustment Type";

        public ReleaseThirteenthMonthPay(DateTime dateFrom, DateTime dateTo, int? currentPeriodID, int organizationId, int userId)
        {
            var data = GetThirteenthMonthPays(dateFrom, dateTo);

            using (var context = new PayrollContext())
            {
                var paystubs = (from p in context.Paystubs.Include(p => p.ActualAdjustments)
                                where p.PayPeriodID.Value == currentPeriodID.Value
                                select p).ToList();

                var product = (from p in context.Products
                               where p.OrganizationID.Value == _organizationId &&
                                    p.PartNo == thirteenthMonthPayname &&
                                    p.Category == adjustmentCategoryName
                               select p).FirstOrDefault();

                var adjustmentCategoryId = context.Products.
                                            Where(p => p.OrganizationID.Value == _organizationId).
                                            Where(p => p.Category == adjustmentCategoryName).
                                            FirstOrDefault().CategoryID;

                // Provide this kind of Product if doesn't exists
                if (product == null)
                {
                    var thirteentMonthPayProduct = new Product()
                    {
                        OrganizationID = _organizationId,
                        Created = DateTime.Now.Date,
                        CreatedBy = _userId,
                        LastUpd = DateTime.Now.Date,
                        LastUpdBy = _userId,
                        PartNo = thirteenthMonthPayname,
                        Name = thirteenthMonthPayname,
                        Description = string.Empty,
                        Category = adjustmentCategoryName,
                        CategoryID = adjustmentCategoryId.Value
                    };
                    var dfsdf = context.Products.Add(thirteentMonthPayProduct);
                    context.SaveChanges();

                    product = thirteentMonthPayProduct;
                }

                foreach (var paystub in paystubs)
                {
                    // Preserve the existing ActualAdjustments, except the 13th month pay
                    var non13thMonthAdjs = paystub.ActualAdjustments.Where(aa => aa.ProductID.Value != product.RowID.Value);

                    // Reset the ActualAdjustments
                    paystub.ActualAdjustments.Clear();

                    // Restore the preserved ActualAdjustments
                    if (non13thMonthAdjs.Any())
                    {
                        foreach (var aa in non13thMonthAdjs)
                            paystub.ActualAdjustments.Add(aa);
                    }

                    // Get first the total thirteenth month pay that the employee will receive.
                    var row = data.Select($"EmployeeID = '{paystub.EmployeeID}'").FirstOrDefault();
                    var thirteenthMonthPay = Convert.ToDecimal(row["ThirteenthMonthPay"]);

                    // Add the new thirteenth pay as an adjustment to the paystub
                    paystub.ActualAdjustments.Add(new ActualAdjustment()
                    {
                        ProductID = product.RowID,
                        OrganizationID = _organizationId,
                        CreatedBy = _userId,
                        Amount = AccuMath.CommercialRound(thirteenthMonthPay)
                    });
                }

                context.SaveChanges();
            }

            this._organizationId = organizationId;
            this._userId = userId;
        }

        private DataTable GetThirteenthMonthPays(DateTime dateFrom, DateTime dateTo)
        {
            var thirteenthMonthPay = $@"
            SELECT pst.EmployeeID, SUM(tmp.Amount) AS ThirteenthMonthPay
            FROM thirteenthmonthpay tmp
            INNER JOIN paystub pst
                    ON pst.RowID = tmp.PaystubID
            WHERE @PayFromDate <= pst.PayFromDate AND
                  pst.PayToDate <= @PayToDate AND
                  pst.OrganizationID = @OrganizationID
            GROUP BY pst.EmployeeID;";

            var @params = new Dictionary<string, object>()
            {
                {"@OrganizationID",_organizationId},
                {"@PayFromDate",dateFrom},
                {"@PayToDate",dateTo}
            };

            //var query = new SqlToDataTable(thirteenthMonthPay, @params);
            //return query.Read();
            return null;
        }
    }
}